### PR TITLE
Miner self-optimizes and prints more neatly

### PR DESCRIPTION
**Pull request by JoshVorick**
When run, the miner will experiment with different global_item_sizes (essentially number of threads) and search for a near optimal number of threads. Each run may result in a different number of threads (e.g. 2048 one run and 4096 the next) because of some variance in speed from call to call. This barely impacts performance because 4096 threads is like calling 2048 threads twice.

The program then tunes the work done in each thread such that every call takes about 3 seconds.

The program prints the current hashrate and the number of blocks mined, updating this line every 3 seconds instead of printing a new line every 3 seconds. The only extra lines that get printed are when the program is searching for the 'optimal' thread count.

### CODE REVIEW

File: gpu-miner.c [hunk 1]
```diff
@@ -95,15 +76,17 @@ int main() {
...
+	if (ret != CL_SUCCESS) { printf("failed to create blockHeadermobj buffer: %d\n", ret); exit(1); }
 	headerHashmobj = clCreateBuffer(context, CL_MEM_READ_WRITE, 32 * sizeof(uint8_t), NULL, &ret);
-	if (ret != CL_SUCCESS) { printf("failed to create targmobj buffer: %d\n", ret); return -1; }
+	if (ret != CL_SUCCESS) { printf("failed to create targmobj buffer: %d\n", ret); exit(1); }
 	targmobj = clCreateBuffer(context, CL_MEM_READ_WRITE, 32 * sizeof(uint8_t), NULL, &ret);
-	if (ret != CL_SUCCESS) { printf("failed to create targmobj buffer: %d\n", ret); return -1; }
+	if (ret != CL_SUCCESS) { printf("failed to create targmobj buffer: %d\n", ret); exit(1); }
 	nonceOutmobj = clCreateBuffer(context, CL_MEM_READ_WRITE, 8 * sizeof(uint8_t), NULL, &ret);
-	if (ret != CL_SUCCESS) { printf("failed to create nonceOutmobj buffer: %d\n", ret); return -1; }
+	if (ret != CL_SUCCESS) { printf("failed to create nonceOutmobj buffer: %d\n", ret); exit(1); }
```
Comment (DavidVorick) [inline gpu-miner.c:159]: why did you change the return to exit(1)? Is that best practice?

Comment (DavidVorick) [inline gpu-miner.c:159]: I am asking b/c I am not sure, not because I want you to change it back.

File: gpu-miner.c [hunk 2]
```diff
@@ -148,91 +131,86 @@ int main() {
...
+	if (ret != CL_SUCCESS) { printf("failed to set fourth kernel arg: \n"); exit(1); }
+	ret = clSetKernelArg(kernel, 5, sizeof(cl_mem), (void *)&iter_per_threadmobj);
+	if (ret != CL_SUCCESS) { printf("failed to set fourth kernel arg: \n"); exit(1); }
+
+	// Rough scan for 'optimal' thread count
+	double hash_rate, prev_hash_rate = 0;
+	global_item_size = 16;
+	iter_per_thread = 16 * 256;
+	while(1) {
+		global_item_size *= 2;
```
Comment (DavidVorick) [inline gpu-miner.c]: you just double every time? Is a more fine-grained search better?

I'd also be curious if iter_per_thread made a difference.

Comment (DavidVorick) [inline gpu-miner.c]: oh I'm dumb ignore me

File: gpu-miner.c [hunk 3]
```diff
@@ -148,91 +131,86 @@ int main() {
...
+			break;
+		}
+		prev_hash_rate = hash_rate;
+	}
+	printf("\rRough search found %zd threads to be the best at %.3f MH/s\n", global_item_size/2, prev_hash_rate);
+	fflush(stdout);
+
+	// Now we know the optimal is betweem global_item_size and global_item_size / 2
+	// Scan intermediate 16 values and pick the highest
+	int dec = (global_item_size - global_item_size / 2) / 16;
```
Comment (DavidVorick) [inline gpu-miner.c]: I don't consider 'dec' to be an immediately obvious variable name, I would prefer step_size or something similar

File: gpu-miner.c [hunk 4]
```diff
@@ -148,91 +131,86 @@ int main() {
...
+	double best_hash_rate = prev_hash_rate;
+	size_t best_item_size = global_item_size / 2;
+	for (i = 0; i < 16; i++) {
+		global_item_size -= dec;
+		double temp = grindNonces(global_item_size, iter_per_thread);
+		while (temp == -1) {
+			// Repeat until no block is found
```
Comment (DavidVorick) [inline gpu-miner.c:259]: lol

File: gpu-miner.c [hunk 5]
```diff
@@ -148,91 +131,86 @@ int main() {
...
+	printf("\rFine search found %zd threads to be the best at %.3f MH/s\n", global_item_size, best_hash_rate);
+	fflush(stdout);
 
-	// Mine blocks until program is interrupted
-	// Each iteration of the loop should take 1-3 seconds
+	// Make each iteration take about 3 seconds
+	clock_t startTime = clock();
+	grindNonces(global_item_size, iter_per_thread);
+	double run_time_seconds = (double)(clock() - startTime) / CLOCKS_PER_SEC;
+	iter_per_thread *= 3 / run_time_seconds;
```
Comment (DavidVorick) [inline gpu-miner.c:280]: interesting that you would tune this live.
1. are you sure it doesn't affect hashrate?
2. is the big worry that the clockrate will change and slow things down? Is there another worry.

Ultimately it seems like a good idea but ' 3 / run_time_seconds' seems weird. You end up with non-even iteration counts.

File: gpu-miner.c [hunk 6]
```diff
@@ -252,3 +232,74 @@ int main() {
  
 	return 0;
 }
+
+// Perform global_item_size * iter_per_thread hashes
+// Return -1 if a block is found
+// Else return the hashrate in MH/s
+double grindNonces(size_t global_item_size, size_t iter_per_thread) {
```
Comment (DavidVorick) [inline gpu-miner.c]: typically the helper functions are put at the top of the file, at least in Nebulous projects anyway.

File: gpu-miner.c [hunk 7]
```diff
@@ -252,3 +232,74 @@ int main() {
...
+	uint8_t blockHeader[80];
+	uint8_t headerHash[32];
+	uint8_t target[32];
+	uint8_t nonceOut[8]; // This is where the nonce that gets a low enough hash will be stored
+	uint8_t nonceOutLock = 0;
+
+	int i;
+	for (i = 0; i < 8; i++)
+		nonceOut[i] = 0;
```
Comment (DavidVorick) [inline gpu-miner.c]: braces please.

```
for {
   doOneThing();
}
```

Comment (DavidVorick) [inline gpu-miner.c]: same for all the other stuff. if, while, etc, always should have braces.

File: gpu-miner.c (hunk 1 above)
Comment (JoshVorick) [inline gpu-miner.c:159]: I believe it is best practice. By using exit(), it's obvious you are terminating the whole program. By using return, it's less clear because return is usually used for things that don't involve terminating the program.

File: gpu-miner.c (hunk 5 above)
Comment (JoshVorick) [inline gpu-miner.c:280]: It might marginally affect hash rate (a bigger iter_per_thread means less time doing non-hashing), but chaning this doesn't affect the 'optimal' thread count. The optimal thread count will be the same no matter the work per thread.

I'm not sure how *= works (does it convert the RHS to int, then multiply or multiply, then convert to int?) but either way, the kernel will do the equivalent of flooring iter_per_thread to the nearest multiple of 256.

I picked 3 seconds somewhat arbitrarily, I can change this if you think 5 seconds between get_block_for_work() calls is okay.

Comment (DavidVorick) [inline gpu-miner.c:280]: 3 is fine.

**Comment by JoshVorick** [comment]
Okay, I made all the changes you recommended. I also updated the 'optimal' thread_num finder. It actually works a lot better now (gets my laptop to 14-16MH/s instead of 12-13MH/s), which is a little funny because I didn't even know my laptop could do that many MH/s.
